src/app/upgrade.cpp

```diff
@@ -0,0 +1,376 @@
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "upgrade.h"
+
+#include <QCoreApplication>
+#include <QDir>
+#include <QRegularExpression>
+#include <QString>
+#include <QSqlDatabase>
+#include <QSqlError>
+#include <QSqlQuery>
+
+#ifndef DISABLE_GUI
+#include <QMessageBox>
+#endif
+
+#include <libtorrent/bdecode.hpp>
+#include <libtorrent/bencode.hpp>
+#include <libtorrent/entry.hpp>
+#include <libtorrent/error_code.hpp>
+#include <libtorrent/version.hpp>
+
+#include "base/bittorrent/db.h"
+#include "base/bittorrent/magneturi.h"
+#include "base/exceptions.h"
+#include "base/global.h"
+#include "base/logger.h"
+#include "base/profile.h"
+#include "base/settingsstorage.h"
+#include "base/utils/fs.h"
+#include "base/utils/misc.h"
+
+namespace
+{
+    bool readFile(const QString &path, QByteArray &buf)
+    {
+        QFile file(path);
+        if (!file.open(QIODevice::ReadOnly)) {
+            qDebug("Cannot read file %s: %s", qUtf8Printable(path), qUtf8Printable(file.errorString()));
+            return false;
+        }
+
+        buf = file.readAll();
+        return true;
+    }
+
+    bool loadTorrentParams(const QByteArray &data, int &prio, QByteArray &magnetUri)
+    {
+        namespace libt = libtorrent;
+
+        libt::error_code ec;
+        libt::bdecode_node fast;
+        libt::bdecode(data.constData(), data.constData() + data.size(), fast, ec);
+        if (ec || (fast.type() != libt::bdecode_node::dict_t)) return false;
+
+        magnetUri = QByteArray::fromStdString(fast.dict_find_string_value("qBt-magnetUri"));
+        prio = fast.dict_find_int_value("qBt-queuePosition");
+
+        return true;
+    }
+
+    void prepareTorrentData(const QSqlQuery &query, int &queuePos, QByteArray &metadata,
+                            QByteArray &fastresume, bool &isMagnetUri)
+    {
+        metadata = query.value(1).toByteArray();
+        tmpFastresume = query.value(2).toByteArray();
+
+        namespace libt = libtorrent;
+
+        libt::error_code ec;
+        libt::bdecode_node fast;
+        libt::bdecode(tmpFastresume.constData(), tmpFastresume.constData() + tmpFastresume.size(), fast, ec);
+        if (ec || (fast.type() != libt::bdecode_node::dict_t)) return;
+
+        libt::entry resumeData(libt::entry::dictionary_t);
+        resumeData = fast;
+
+        resumeData["qBt-savePath"] = query.value(4).toString().toStdString();
```

Comment:
Nothing inside this method indicates the layout of the query result. It would be better to use field names instead of indexes.

Reply:
@glassez:  **Off-Topic**:  The plural form of "index" is "indices".

Reply:
Aren't both forms acceptable?
https://grammarist.com/usage/indexes-indices/

Reply:
@glassez:  From the same link (emphasis mine): 

>Meanwhile, **indices is generally preferred in mathematical, financial, and technical contexts**, while indexes is relatively common in general usage.

Reply:
OK, thanks.

Reply:
> It would be better to use field names instead of indexes.

Unforunately they don't correlate one-to-one. The index numbers are dependent on what you asked with your `SELECT` statement.

Reply:
>The index numbers are dependent on what you asked with your SELECT statement.

Yes. But I can't see what you asked with your SELECT statement since it is exist in somewhere else. But the field names speak for themselves (I mean field/column names from SQL table).

Reply:
>(I mean field/column names from SQL table).

`QSqlQuery::value(const QString &name)` the docs say that this method is inefficient.

Bear in mind that `prepareTorrentData()` is a helper function called from one place only. So the reader/reviewer can backtrack a little bit. Is this sufficient to not seek out a more rigid solution?
How about if I put comments to indicate which column each number represents?

Reply:
>How about if I put comments to indicate which column each number represents?

This is more or less clear as it is. There's another problem. Do the indexes used match the correct columns? For example, you changed the query but forgot to change the indexes.

Reply:
>Bear in mind that `prepareTorrentData()` is a helper function called from one place only.

I wonder why you're passing QSqlQuery here. Why not extract all needed data in outer method and pass it here as dictionary?